src/database/database.cpp

```diff
@@ -0,0 +1,1099 @@
+ * Private
+ **********************************************************************************************************************/
+
+RetWithError<bool> Database::TableExist(const std::string& tableName)
+{
+    size_t count {0};
+
+    try {
+        Poco::Data::Statement statement {*mSession};
+
+        statement << "SELECT count(*) FROM sqlite_master WHERE name = ? and type='table'", bind(tableName), into(count);
+
+        if (statement.execute() == 0) {
+            return {false, ErrorEnum::eNotFound};
+        }
+
+    } catch (const Poco::Exception& e) {
+        return {false, Error(AOS_ERROR_WRAP(ErrorEnum::eFailed), e.what())};
+    }
+
+    return {count > 0, ErrorEnum::eNone};
+}
+
+Error Database::DropAllTables()
+{
+    try {
+        LOG_WRN() << "Dropping all tables";
+
+        *mSession << "DROP TABLE IF EXISTS config;", now;
+        *mSession << "DROP TABLE IF EXISTS network;", now;
+        *mSession << "DROP TABLE IF EXISTS services;", now;
+        *mSession << "DROP TABLE IF EXISTS trafficmonitor;", now;
+        *mSession << "DROP TABLE IF EXISTS layers;", now;
+        *mSession << "DROP TABLE IF EXISTS instances;", now;
+
+    } catch (const Poco::Exception& e) {
+        return Error(AOS_ERROR_WRAP(ErrorEnum::eFailed), e.what());
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Database::CreateConfigTable()
+{
+    auto [tableExists, err] = TableExist("config");
+
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    if (tableExists) {
+        uint64_t operationVersion {0};
+
+        Tie(operationVersion, err) = GetOperationVersion();
+        if (!err.IsNone()) {
+            return err;
+        }
+
+        if (operationVersion == sm::launcher::Launcher::cOperationVersion) {
+            return ErrorEnum::eNone;
+        }
+
+        if (err = DropAllTables(); !err.IsNone()) {
+            return Error(AOS_ERROR_WRAP(err), "failed to drop all tables");
+        }
+    }
+
+    *mSession << "CREATE TABLE config ("
+                 "operationVersion INTEGER, "
+                 "cursor TEXT, "
+                 "envvars TEXT, "
+                 "onlineTime TIMESTAMP);",
+        now;
+
+    try {
+        *mSession << "INSERT INTO config ("
+                     "operationVersion, "
+                     "onlineTime) values(?, ?);",
+            bind(Poco::Tuple<uint32_t, uint64_t>(sm::launcher::Launcher::cOperationVersion, Time::Now().UnixNano())),
+            now;
+    } catch (const Poco::Exception& e) {
+        return AOS_ERROR_WRAP(ErrorEnum::eFailed);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+void Database::CreateTables()
+{
+    *mSession << "CREATE TABLE IF NOT EXISTS network ("
+                 "networkID TEXT NOT NULL PRIMARY KEY, "
+                 "ip TEXT, "
+                 "subnet TEXT, "
+                 "vlanID INTEGER, "
+                 "vlanIfName TEXT);",
+        now;
+
+    *mSession << "CREATE TABLE IF NOT EXISTS services ("
+                 "id TEXT NOT NULL , "
```

Comment:
id and version are key fields

Reply:
It's declared "PRIMARY KEY(id, version));" a few lines below.